File: 2020_2학기_알고리즘/15주차_기말/1.c
#include <stdio.h>

int n,m;
int size[1001] = {0, };
int value[1001] = {0, };
int count[1001] = {0, };
int dp[1001][1001][100000] = {0, };

int func(int i, int s, int v){
	if (i >= m) return 0;
	if (dp[i][s][v]) return dp[i][s][v];
	int a = func(i+1, s,v);
	int b = 0;
	if (count[i] > 0 && s + size[i] <= n){
		count[i]--;
		b = value[i]+func(i,s+size[i], v+value[i]);
	}
	dp[i][s][v] = a > b ? a : b;
	return dp[i][s][v];
}

int main(){
	scanf("%d",&n);
	scanf("%d",&m);
	for (int i = 0; i < m; i++)
		scanf("%d",&size[i]);
	for (int i = 0; i < m; i++)
		scanf("%d",&value[i]);
	for (int i = 0; i < m; i++)
		scanf("%d",&count[i]);
	printf("%d\n",func(0,0,0));
}

File: 2020_2학기_알고리즘/15주차_기말/1.py
import sys

n = int(sys.stdin.readline())
m = int(sys.stdin.readline())

size = list(map(int, sys.stdin.readline().split(' ')))
money = list(map(int, sys.stdin.readline().split(' ')))
count = list(map(int, sys.stdin.readline().split(' ')))

dp1 = [[0 for _ in range(m+1)] for _ in range(n+1)]
dp = [dp1 for _ in range(10000)]

def func(i, s, value):
	if i >= m:
		return 0
	if dp[i][s][value] > 0:
		return dp[i][s][value]
	a = func(i+1,s,value)
	b = 0
	if count[i] > 0 and s + size[i] <= n:
		count[i] -= 1
		b = money[i] + func(i, s+size[i], value + money[i])
	if a > b :
		dp[i][s][value] = a
	else:
		dp[i][s][value] = b
	return dp[i][s][value]
print(dp[4][10][9999])
print(func(0,0,0))
